File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/BcComponentsException.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.dcbor.CborException

/**
 * Errors that can occur when using the bc-components library.
 *
 * This sealed class hierarchy models the various error conditions from the
 * Rust `bc_components::Error` enum. Only the error variants needed by the
 * default-features subset are included; additional variants can be added as
 * later translation units are completed.
 */
sealed class BcComponentsException(
    message: String,
    cause: Throwable? = null,
) : Exception(message, cause) {

    /** Invalid data size for the specified type. */
    class InvalidSize(
        val dataType: String,
        val expected: Int,
        val actual: Int,
    ) : BcComponentsException(
        "invalid $dataType size: expected $expected, got $actual",
    )

    /** Invalid data format or content. */
    class InvalidData(
        val dataType: String,
        val reason: String,
    ) : BcComponentsException("invalid $dataType: $reason")

    /** Data too short for the specified type. */
    class DataTooShort(
        val dataType: String,
        val minimum: Int,
        val actual: Int,
    ) : BcComponentsException(
        "data too short: $dataType expected at least $minimum, got $actual",
    )

    /** Cryptographic operation failed. */
    class Crypto(val msg: String) : BcComponentsException(
        "cryptographic operation failed: $msg",
    )

    /** CBOR encoding or decoding error. */
    class Cbor(val error: CborException) : BcComponentsException(
        "CBOR error: ${error.message}",
        error,
    )

    /** SSKR error. */
    class Sskr(val error: com.blockchaincommons.sskr.SskrException) :
        BcComponentsException("SSKR error: ${error.message}", error)

    /** Data compression/decompression failed. */
    class Compression(val msg: String) : BcComponentsException(
        "compression error: $msg",
    )

    /** Post-quantum cryptographic operation failed. */
    class PostQuantum(val msg: String) : BcComponentsException(
        "post-quantum error: $msg",
    )

    /** Security level mismatch between key and signature/ciphertext. */
    class LevelMismatch : BcComponentsException(
        "security level mismatch between key and signature/ciphertext",
    )

    /** General error with custom message. */
    class General(val msg: String) : BcComponentsException(msg)

    companion object {
        fun general(msg: String): BcComponentsException = General(msg)

        fun invalidSize(dataType: String, expected: Int, actual: Int): BcComponentsException =
            InvalidSize(dataType, expected, actual)

        fun invalidData(dataType: String, reason: String): BcComponentsException =
            InvalidData(dataType, reason)

        fun dataTooShort(dataType: String, minimum: Int, actual: Int): BcComponentsException =
            DataTooShort(dataType, minimum, actual)

        fun crypto(msg: String): BcComponentsException = Crypto(msg)

        fun compression(msg: String): BcComponentsException = Compression(msg)

        fun postQuantum(msg: String): BcComponentsException = PostQuantum(msg)

        fun levelMismatch(): BcComponentsException = LevelMismatch()
    }
}

/**
 * Converts a [BcComponentsException] into a [CborException] for interop
 * with dCBOR interface implementations.
 */
fun BcComponentsException.toCborException(): CborException = when (this) {
    is BcComponentsException.Cbor -> error
    else -> CborException.msg(message ?: "unknown error")
}

File: csharp/BCComponents/BCComponents/BCComponentsException.cs
namespace BlockchainCommons.BCComponents;

/// <summary>
/// Exception type for bc-components errors.
/// Provides factory methods for common error categories such as invalid sizes,
/// invalid data, cryptographic failures, SSH errors, and more.
/// </summary>
public sealed class BCComponentsException : Exception
{
    /// <summary>
    /// Initializes a new instance of <see cref="BCComponentsException"/> with the specified message.
    /// </summary>
    /// <param name="message">The error message.</param>
    public BCComponentsException(string message) : base(message) { }

    /// <summary>
    /// Initializes a new instance of <see cref="BCComponentsException"/> with the specified message
    /// and inner exception.
    /// </summary>
    /// <param name="message">The error message.</param>
    /// <param name="innerException">The inner exception.</param>
    public BCComponentsException(string message, Exception innerException) : base(message, innerException) { }

    /// <summary>Creates an error for an invalid data size.</summary>
    /// <param name="dataType">The name of the data type.</param>
    /// <param name="expected">The expected size.</param>
    /// <param name="actual">The actual size received.</param>
    /// <returns>A new <see cref="BCComponentsException"/>.</returns>
    public static BCComponentsException InvalidSize(string dataType, int expected, int actual)
    {
        return new BCComponentsException($"invalid {dataType} size: expected {expected}, got {actual}");
    }

    /// <summary>Creates an error for invalid data content.</summary>
    /// <param name="dataType">The name of the data type.</param>
    /// <param name="reason">The reason the data is invalid.</param>
    /// <returns>A new <see cref="BCComponentsException"/>.</returns>
    public static BCComponentsException InvalidData(string dataType, string reason)
    {
        return new BCComponentsException($"invalid {dataType}: {reason}");
    }

    /// <summary>Creates an error when data is too short.</summary>
    /// <param name="dataType">The name of the data type.</param>
    /// <param name="minimum">The minimum required size.</param>
    /// <param name="actual">The actual size received.</param>
    /// <returns>A new <see cref="BCComponentsException"/>.</returns>
    public static BCComponentsException DataTooShort(string dataType, int minimum, int actual)
    {
        return new BCComponentsException($"data too short: {dataType} expected at least {minimum}, got {actual}");
    }

    /// <summary>Creates an error for a cryptographic operation failure.</summary>
    /// <param name="message">A description of the failure.</param>
    /// <returns>A new <see cref="BCComponentsException"/>.</returns>
    public static BCComponentsException Crypto(string message)
    {
        return new BCComponentsException($"cryptographic operation failed: {message}");
    }

    /// <summary>Creates an error for an SSH operation failure.</summary>
    /// <param name="message">A description of the failure.</param>
    /// <returns>A new <see cref="BCComponentsException"/>.</returns>
    public static BCComponentsException Ssh(string message)
    {
        return new BCComponentsException($"SSH operation failed: {message}");
    }

    /// <summary>Creates an error for a data compression/decompression failure.</summary>
    /// <param name="message">A description of the failure.</param>
    /// <returns>A new <see cref="BCComponentsException"/>.</returns>
    public static BCComponentsException Compression(string message)
    {
        return new BCComponentsException($"compression error: {message}");
    }

    /// <summary>Creates an error for a post-quantum cryptography failure.</summary>
    /// <param name="message">A description of the failure.</param>
    /// <returns>A new <see cref="BCComponentsException"/>.</returns>
    public static BCComponentsException PostQuantum(string message)
    {
        return new BCComponentsException($"post-quantum cryptography error: {message}");
    }

    /// <summary>Creates an error for an SSH agent operation failure.</summary>
    /// <param name="message">A description of the failure.</param>
    /// <returns>A new <see cref="BCComponentsException"/>.</returns>
    public static BCComponentsException SshAgent(string message)
    {
        return new BCComponentsException($"SSH agent error: {message}");
    }

    /// <summary>Creates a general-purpose error with a custom message.</summary>
    /// <param name="message">The error message.</param>
    /// <returns>A new <see cref="BCComponentsException"/>.</returns>
    public static BCComponentsException General(string message)
    {
        return new BCComponentsException(message);
    }

    /// <summary>Creates an error for a signature level mismatch.</summary>
    /// <returns>A new <see cref="BCComponentsException"/>.</returns>
    public static BCComponentsException LevelMismatch()
    {
        return new BCComponentsException("signature level does not match key level");
    }
}
